src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageEndpoint.cs

```diff
@@ -298,5 +298,96 @@ public async Task<RazorMapToDocumentEditsResponse> Handle(RazorMapToDocumentEdit
                 HostDocumentVersion = documentVersion,
             };
         }
+
+        public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var unmappedDiagnostics = request.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
```

Comment:
Should just ToARray this here to prevent iterations re-calculating the filtered list each time. Would also make your other loops below a little simpler